### ISSUE TITLE
NNP Factory Class Function Redesign

### ISSUE THREAD

**Issue by MarshallYan** [choderalab/modelforge#269]
Based on my discussion with @wiederm and @chrisiacovella, it might be beneficial to redesign the logic of this middle layer "factory" and rename anything that is currently named as "model".

Note: We should explicitly call a model a "trainer" when it is a `ModelTrainer` (the class itself should be `PotentialTrainer`) object in training mode, and a "potential" in inference mode when it is a `Potential` object.

The general idea is that generate a potential **trainer** or loading a potential **trainer** from either a checkpoint or a Pytorch state_dict should be an attribute of the `ModelTrainer` class. To load a potential for inference, on the other hand can be implemented by a) a new function in the `Potential` class, b) a new class named as `PotentialInferrer` which holds those JAX converter, etc, or c) a function directly placed in "models.py" (we should call it "potential.py" respectively). *I personally prefer the b) option.*

In this design:
  1. There won't be confusion of what a model is.
  2. All trainer properties are under its class, so when a user tries to initialize a trainer, there's no need to jump between "models.py" (will be "potential.py") and "training.py".
  3. Anything about inference will be in "potential.py" since you will use a **potential** model to do inference and should have nothing to do with "training.py".
  4. Upload model will be implemented under a Trainer class as there's no reason to upload models when they're used only for inference.
  5. Download models will be implemented seperately for training and inference. A user will have to choose to download a a) checkpoint file, b) a state_dict file, or c) an entire model file ("potential.pt", we need to further whether is necessary at any situation). This way, we can easily set a naming convention (or separate directories holding different types of potential models) so that users will only download inference potential for inference and vice versa.

**Comment by chrisiacovella** [comment]
To elaborate a little more, this would involve removing the `NeuralNetworkPotentialFactory`. This class is basically just to switch between setting up a `ModelTrainer` instance for training, or return an instance of a `Potential` for inference.  We can instantiate the `ModelTrainer` directly from, e.g., `read_config_and_train`.  We would probably want another wrapper as Shuai mentioned, because we do have some extra logic needed related to exporting to pytorch or converting to jax. 

Regarding the statedict, I think we can improve the saving and loading by a refactor.  In short, the code has a class called `ComputeInteractingAtomPairs` which instantiates the training neighbor list inside it.  This results in additional hierarchy (as cutoff is registered inside the neighbor list, not in the `ComputeInteractingAtomPairs` class).  I don't think we need these classes to be separate and instead should just merge them into a single Neighborlist class.

**Comment by wiederm** [comment]
We currently have the following class structure:
- `class Potential(torch.nn.Module)` is used for inference, it can be exported to JAX and is torchscriptable. It is implemented in `modelfoge.potential.models`.
- `class TrainingAdapter(pytorch.Lightning)` is a Pytorch Lightning module that sets up the Potential class inside its `__init__` call and contains all training-related methods (e.g. methods to perform a training/validation step, torchmetrics, optimizer). It also saves the hyperparameters to reproduce the exact same `TrainingAdapter` state from a checkpoint file (including the training state). It is implemented in `modelforge.train.training`
- `class ModelTrainer`, which contains the `TrainingAdapter`. This class performs all the setup to initialize a training (callbacks, loggers, datamodule, Trainer) and, with the `train_potential` methods, also contains the method that performs the training.

The main idea was that for inference mode, we only want the bareback `Potential` class, for training, we want to wrap all the training **setup** logic inside a class (including dataset splitting, callback/Trainer/etc setup) and wrap the Potential in a pytorch.Lightning class so that we also define how the training **logic** looks like in a class structure for reusability.

In that way, the only thing that we need to do to perform training is `ModelTrainer().train_potential()` and we perform a fitting/validation/testing loop and get back a `pL.Trainer` class that can be used further.

**Comment by wiederm** [comment]
I agree with @MarshallYan in general, there are a few low hanging improvements that would make this easier:
- use potential instead of model in variable names (class names already reflect that)
- use trainer instead of model when it's a `ModelTrainer`

ad 2. and 3.: I believe this is already the case. All trainer properties are in training.py (and all loss functionality is in loss.py), and all inference functionality is in model.py (or potential.py)

ad 4. checkpoints (and model states) are automatically logged as W&B artifacts based on validation set performance during training, I don't think there is anything else to do.

ad 5: this is up for debate I think. I'd have suggested that we use a checkpoint file in all use cases, for inference we extract the state_dict, for training we load the full checkpoint file.

**Comment by chrisiacovella** [comment]
Let us try to avoid doing all of this in one giant PR, instead breaking it up into smaller pieces that will be easier to review/discuss:

1. PR only addressing the change in naming convention
2. PR restructuring the NNP factory (i.e., splitting contents up)
3. PR restructuring neighbor list to avoid having to massage the loading of the state_dict (this can be done independent of the other 2 PRs).
4. PR addressing how to best save/upload/fetch/load trained parameters from wandb.
